KFSPTS-25884 create CSV file for jaggaer upload batch job

jhulslander opened the pull request:
An initial review can be started, but please do not merge this yet.  I have made some code changes based on feedback I have gotten, but I need to do further local validation.
Also, I will need to re-base this PR and incorporate the FIPS to ISO service.

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -57,16 +67,137 @@ protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<Jaggaer
+    }
+    
+    protected String[] buildPartyHeader() {
+        String[] record = { "PARTY", "OverrideDupError", "ERPNumber", "SciQuestID", "ContractPartyName", "DoingBusinessAs",
```
Would it be better to put these header labels in a constants class or enum for convenience? The same goes for the other types of headers.

jhulslander replied:
OK, moved to an enum

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -187,5 +212,151 @@ static Stream<Arguments> jaggaerContractPartyUploadDtoToStringParameters() {
+    void testBuilderVendorCSVRowArray(JaggaerContractPartyUploadDto vendorDto, String[] expectedCsvData) {
+
+        String[] actualCsvData = jaggaerGenerateContractPartyCsvServiceImpl.builderVendorCSVRowArray(vendorDto);
+        assertEquals(16, expectedCsvData.length);
```
Depending on what changes are made to the reorganization of the header labels, these numeric literals could be replaced by array length or collection size calls. The same goes for the similar situation in the address-row-testing method.

jhulslander replied:
OK

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -187,5 +212,151 @@ static Stream<Arguments> jaggaerContractPartyUploadDtoToStringParameters() {
+        }
+    }
+
+    static Stream<Arguments> builderVendorCSVRowArrayParameters() {
```
There are various test methods and arg-building methods that say "builder" instead of "build". Is that the way you meant to name them?

jhulslander replied:
yes, switching to buildXXParameters

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -26,35 +27,59 @@
+    private static final String VENDOR3_OTHER_NAME = "other name";
+    private static final String VENDOR3_DBA = "Acme";
+    private static final String VENDOR3_CONTRACT_PARTY_NAME = "Acme Inc";
+    private static final String VENDOR3_ERP_NUMBER = "123456-0";
```
If you end up adding even more vendors or addresses to test with, you should consider creating a helper enum or class to simplify managing the test data. Such a setup may also help with creating test cases that validate writing the CSV file to disk, which would be good test cases to add if you can.

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -187,5 +212,151 @@ static Stream<Arguments> jaggaerContractPartyUploadDtoToStringParameters() {
+        assertEquals(expectedContactHeader, actualContactHeader);
+    }
+
+    private String printArrayCommaDelim(String[] data) {
+        StringBuilder sb = new StringBuilder();
```
You could simplify this concatenation by using one of the StringUtils.join() methods, or by using a Stream and using Collectors.joining() to build the concatenated string.

jhulslander replied:
ok

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -114,6 +118,7 @@ private String buildVendorAddressSql(JaggaerContractUploadProcessingMode process
         sb.append(buildJoinClauseWithPOVendors(true));
         sb.append(buildRestrictActiveRows(true));
         sb.append(buildTimeFrameRestrictionClause(processingMode, processingDate));
+        sb.append("AND VA.VNDR_ADDR_TYP_CD in ('RM', 'PO') ");
```
You may want to create a follow-up user story to make the SQL building a little cleaner and to create it as a parameterized query. Some previous PR(s) introduced some helper classes (like CuSqlQuery and CuSqlChunk) that make it simpler to build a parameterized SQL query.

jhulslander replied:
I am not sure if making this query parameterized is ideal.  I like being able to log the SQL being executed to help debug any problems. Parameterizing SQL statements is a nice way of preventing user input SQL injection.  We don't have user input here.
If you think there is an alternative way of creating this SQL that is simpler, please let me know and I can create a follow up user story to adjust the SQL later.

cah292 replied:
For the example line given here, it's okay to use the hard-coded SQL literals for now, though ideally it would be preferable to use constants to store the values and then pass them in as query parameters. However, I've noticed other areas in this DAO class where the code is appending a String-formatted date into the SQL, and is using a SQL TO_DATE function to convert the String back into a Date on the database side. At some point in the future, it would be better to pass such values into the query as DATE or TIMESTAMP JDBC parameters.

The CuSqlQuery and CuSqlChunk helper classes I mentioned earlier could help simplify building this query in a future user story. They make it more convenient to set up parameterized queries, and they also keep track of the query parameters so that both the query string and the parameters can easily be written to the logs as needed. The classes CuSqlQueryTest and DocumentMaintenanceDaoJbc have examples of the capabilities of these helper objects.

jhulslander replied:
I created https://jira.cornell.edu/browse/KFSPTS-26048

cah292 reviewed (changes requested):
Here are my initial thoughts on the CSV-building code.

Also, given the complexity of Jaggaer's CSV header row and content row setup, I agree that your use of the array-based CSV-building methods would work out better than trying to create a "mapping strategy" helper class.

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -62,8 +62,12 @@ public List<JaggaerContractAddressUploadDto> findJaggaerContractAddress(JaggaerC
                 dto.setAddressType(JaggaerAddressType.findJaggaerAddressTypeFromKfsAddressTypeCode(resultSet.getString(FIELD_NAMES.VNDR_ADDR_TYP_CD)));
-                dto.setPrimaryType(resultSet.getString(FIELD_NAMES.VNDR_DFLT_ADDR_IND));
-                dto.setActive(StringUtils.EMPTY);
+                if (StringUtils.equals(resultSet.getString(FIELD_NAMES.VNDR_DFLT_ADDR_IND), "Y")) {
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -22,14 +22,14 @@
                 JaggaerContractPartyUploadDto dto = new JaggaerContractPartyUploadDto();
                 dto.setRowType(JaggaerContractPartyUploadRowType.PARTY);
-                dto.setOverrideDupError(false);
+                dto.setOverrideDupError("False");
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -1,25 +1,35 @@
     private static final Logger LOG = LogManager.getLogger();
     
-    protected JaggaerUploadDao jaggaerUploadDao;
+    private String jaggaerUploadCreationDriectory;
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -57,16 +67,137 @@ protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<Jaggaer
+        
+        List<String[]> csvData = buildDataToPrint(jaggaerUploadDtos);
+
+        try (FileWriter fileWriter = new FileWriter(fullyQualifiedCreationDirectoryFileName);
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -57,16 +67,137 @@ protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<Jaggaer
+    }
+    
+    private String generateCsvOutputFileName(JaggaerContractUploadProcessingMode processingMode) {
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US);
```
ok

jhulslander commented:
I am still doing some local validation, but this is ready for review again

jhulslander commented:
FYI, we need to revert this FIPS - ISO conversion code out of develop, that may cause problems on this PR.  Please do NOT merge this

cah292 commented:
Just to confirm, is the problem with the FIPS-ISO conversion code serious enough to justify reverting it, or does it just need a minor follow-up patch to fix a small issue?

jhulslander commented:
at this point we think it needs reverting

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorFixture.java:
```diff
@@ -0,0 +1,98 @@
+public enum VendorFixture {
+    BASIC_VENDOR("12345-0", "Acme Testing Company", "US", "http://www.google.com"),
+    BASIC_VENDOR_WITH_TAX_ID("98765-0", "Jane Doe Paint Services", "US", "65479", "http://www.yahoo.com"),
+    FULL_VENOOR(JaggaerContractPartyUploadRowType.PARTY, StringUtils.EMPTY, "active", "123456-0",
```
Minor typo in name of enum constant.

jhulslander replied:
ok

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorFixture.java:
```diff
@@ -0,0 +1,98 @@
+    FULL_VENOOR(JaggaerContractPartyUploadRowType.PARTY, StringUtils.EMPTY, "active", "123456-0",
+            CUPurapConstants.FALSE_STRING, "Acme Inc", "Acme", "other name", "US", JaggaerContractPartyType.SUPPLIER,
+            "primary", JaggaerLegalStructure.C_CORPORATION, "foo type", "369852", "vat number", "www.google.com"),
+    FULL_VENOOR_FOR_CSV(JaggaerContractPartyUploadRowType.PARTY, "vendor SciQuest-id", "active", "66666-0",
```
Minor typo in name of enum constant.

jhulslander replied:
ok

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorAddressFixture.java:
```diff
@@ -0,0 +1,97 @@
+            "Apartment 666", "Freeville", "NY", "13068"),
+    BASIC_VENDOR_ADDRESS_2(VendorFixture.BASIC_VENDOR.ERPNumber, "98765", JaggaerAddressType.REMIT, "45 Palm Street",
+            "Apartment 1F", "Ithaca", "NY", "14850"),
+    FULL_VENDOR_FULL_ADDRSS(JaggaerContractPartyUploadRowType.ADDRESS, StringUtils.EMPTY, "Active", VendorFixture.FULL_VENOOR.ERPNumber,
```
Minor typo in name of enum constant.

jhulslander replied:
ok

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorAddressFixture.java:
```diff
@@ -0,0 +1,97 @@
+    FULL_VENDOR_FULL_ADDRSS(JaggaerContractPartyUploadRowType.ADDRESS, StringUtils.EMPTY, "Active", VendorFixture.FULL_VENOOR.ERPNumber,
+            "addressId", "ABCD", JaggaerAddressType.FULFILLMENT, "primary type", "US", "street line 1", "street line 2", "street line 3",
+            "city", "NY", "14850", "6072559900", "18002809900", "fax", StringUtils.EMPTY),
+    FULL_VENDOR_FULL_ADDRSS_FOR_CSV(JaggaerContractPartyUploadRowType.ADDRESS, "vendor address sciquest id", "Active", VendorFixture.FULL_VENOOR_FOR_CSV.ERPNumber,
```
Minor typo in name of enum constant.

jhulslander replied:
ok

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorFixture.java:
```diff
@@ -0,0 +1,98 @@
+        dto.setTaxIdentificationNumber(taxIdentificationNumber);
+        dto.setVATRegistrationNumber(VATRegistrationNumber);
+        dto.setWebsiteURL(websiteURL);
+        ;
```
Unnecessary semicolon on empty line.

jhulslander replied:
ok

cah292 reviewed (changes requested):
Here are my findings after performing a second pass. I still have a little more to review in the unit test code, but I might wait on that until the items below are addressed and the code gets rebased with an updated FIPS-to-ISO fix.

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -176,14 +186,24 @@ private String buildVendorNumber(String vendorHeaderId, String vendorDetailId) {
+        String isoCountry = StringUtils.EMPTY;
+        
+        try {
+            isoCountry = isoFipsConversionService.convertFIPSCountryCodeToActiveISOCountryCode(fipsCountryCode);
+        } catch (RuntimeException runtimeException) {
+            LOG.error("convertToISOCountry, unable to get ISO country for FIPS country " + fipsCountryCode, runtimeException);
```
Yes, that is what the functionals requested.  If the conversion has an error, they want to an empty country

cah292 replied:
Ok, sounds good. Thanks for checking.

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/CUPurapConstants.java:
```diff
@@ -371,4 +383,96 @@ public static JaggaerContractUploadProcessingMode findJaggaerContractUploadProce
+        ACTIVE("Active"),
+        CONTRACT_PARTY_TYPE("ContractPartyType"),
+        PRIMARY("Primary"),
+        LEGALS_TRUCTURE("LegalStructure"),
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -176,14 +186,24 @@ private String buildVendorNumber(String vendorHeaderId, String vendorDetailId) {
+        if (StringUtils.isBlank(fipsCountryCode)) {
             LOG.debug("convertToISOCountry, empty FIPS country found returning US as default");
-            fIPSCountry = "US";
+            fipsCountryCode = "US";
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -57,16 +72,134 @@ protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<Jaggaer
+                .toArray(String[]::new);
+    }
+
+    protected String[] builderVendorCSVRowArray(JaggaerContractPartyUploadDto vendorDto) {
```
yes they should

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -158,34 +102,165 @@ void testGetJaggerContractsDto() {
+    public void validateCsvRowArray(String[] expectedCsvData, String[] actualCsvData) {
+        for (int i = 0; i < expectedCsvData.length; i++) {
+            String actualDatum = actualCsvData[i];
+            String expectedDataum = expectedCsvData[i];
```
ok

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -158,34 +102,165 @@ void testGetJaggerContractsDto() {
+        File actualJaggaerFile = new File(BASE_CSV_DIRECTORY + "JaggaerUpload_found_by_PO_search_19961130_233100.csv");
+        
+        FileUtils.readFileToString(expectedJaggaerFile, StandardCharsets.UTF_8);
+        assertEquals(FileUtils.readFileToString(expectedJaggaerFile, StandardCharsets.UTF_8), 
```
ok

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorAddressFixture.java:
```diff
@@ -0,0 +1,97 @@
+package edu.cornell.kfs.module.purap.batch.service.impl.fixcture;
```
ok

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixcture/VendorAddressFixture.java:
```diff
@@ -0,0 +1,97 @@
+package edu.cornell.kfs.module.purap.batch.service.impl.fixcture;
+
+import org.apache.commons.lang.StringUtils;
```
ok

db483 commented:
retest this please

jhulslander commented:
This is ready for review again.  I've added the conversion service code, fixed a few small bugs, I also added some additional logging to JaggaerUploadDaoJdbc just so the job doesn't appear to be "hung" if someone is monitoring the batch log for the job.  If you think there is a better way to accomplish that, please let me know.

cah292 commented on src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:
```diff
@@ -16,20 +18,27 @@
     public List<JaggaerContractPartyUploadDto> findJaggaerContractParty(JaggaerContractUploadProcessingMode processingMode, String processingDate) {
         try {
             RowMapper<JaggaerContractPartyUploadDto> rowMapper = (resultSet, rowNumber) -> {
+                if (rowNumber == 0 || rowNumber % 100 == 0) {
```
You could simplify this condition to just use the MOD (%) part of the check, since 0 MOD 100 should be zero. The same goes for the similar areas in this file.

jhulslander replied:
OK

cah292 commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:
```diff
@@ -158,34 +103,179 @@ void testGetJaggerContractsDto() {
+        
+        createTargetTestDirectory();
+        
+        jaggaerGenerateContractPartyCsvServiceImpl.generateCsvFile(jaggaerUploadDtos, JaggaerContractUploadProcessingMode.PO);
```
Should this call be placed inside the method's "try" block?

jhulslander replied:
I will move it

cah292 reviewed (commented):
I found a few areas that might need cleanup, but otherwise the changes look good. Since they're largely related to either unit test setup or minor non-unit-test functionality, let me know whether or not you want some or all of them deferred to a future ticket.

There's also one related comment that I accidentally posted earlier than the others; just wanted you to be aware so that you can review that one too. Sorry for the confusion.

jhulslander commented on src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:
```diff
@@ -57,16 +72,134 @@ protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<Jaggaer
+            writer.writeAll(csvData);
+        } catch (IOException e) {
+            LOG.error("generateCsvFile, Error writing to file " + fullyQualifiedCreationDirectoryFileName, e);
+            throw new RuntimeException(e);
```
ok

jhulslander commented on src/main/java/edu/cornell/kfs/sys/CUKFSConstants.java:
```diff
@@ -209,5 +209,7 @@ public static final class WebappPaths {
     public final static boolean MAPPING_ACTIVE = true;
     public final static boolean MAPPING_INACTIVE = false;
+    
+    public static final String COUNTRY_CODE_US = "US";
```
I'll switch

jhulslander commented on src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/fixture/VendorAddressFixture.java:
```diff
@@ -0,0 +1,98 @@
+package edu.cornell.kfs.module.purap.batch.service.impl.fixture;
+
+import org.apache.commons.lang.StringUtils;
```
ok

jhulslander commented:
this is ready for review again

cah292 reviewed (approved):
Changes look good, unit tests pass, and my concerns have been addressed. Merging.
